Asset Pipeline Merge

devshgraphicsprogramming opened the pull request:
Can't merge automatically yet, three macro-scale issues to resolve:
- Merge master into the branch first
- Address the re-raised issues and comments from the last few days
- Update the examples to compile and run under the new APIs

Crisspl committed:
asset-pipeline: Initial interfaces

My comments are marked with '(Criss)' to be easily findable with ctrl+F, please answer on github in form of comments to commit so it's easily searchable

Crisspl committed:
object-cache, asset-pipeline: Rolled back to std containers and added/modified some memb. funcs

Roll back is necessary for now because apparently it's compiler bug that causes msvc to crash while using caches with our containers from core::.
https://social.msdn.microsoft.com/Forums/office/en-US/ea2b805b-5464-4678-af2f-f20331c0367f/visual-studio-enterprise-2015-update-3-compiler-has-stopped-working
Also added clear() member func (needed for IAssetManager::clearAllCaches()) and changed removeObject() to return bool

Crisspl committed:
Added core::array container

Crisspl committed:
asset-pipeline: Some changes to IAsset, IAssetLoader, IAssetWriter

Crisspl committed:
asset-pipeline: IAssetManager initial implementation

Missing some functions, important comments marked with 'todo'

Crisspl committed:
object-cache: Custom allocator

Crisspl committed:
asset-pipeline: Some small changes to IAssetManager

Crisspl committed:
asset-pipeline: Implemented IAsset interface for mesh, meshbuffer and buffer

Crisspl committed:
asset-pipeline: Initial PLY loader adjustments

No LoaderOverride usage yet

Crisspl committed:
asset-pipeline: Inital STL loader adjustments

No LoaderOverride usage yet

Crisspl committed:
asset-pipeline: Made IAssetLoader and IAssetWriter derivatives of IReferenceCounted

+ missing IAssetLoader.cpp

Crisspl committed:
irrelevant commit

i forgot to reset it and didnt manage to drop this

Crisspl committed:
object-cache: fixed clear()

Crisspl committed:
Oops, this must be public base

Crisspl committed:
IAsset::isCached == false by default

Crisspl committed:
asset-pipeline: Added IAssetLoaderOverride::setAssetCacheKey() required at least for internal sub-assets

You'll probably want to talk about this

Crisspl committed:
asset-pipeline: IAsset::isCached is now controlled directly by cache containers (greet/dispose)

Crisspl committed:
asset-pipeline: BAW loader adjustments

Probably missing some LoaderOverride calls

Crisspl committed:
asset-pipeline: IAssetLoaderOverride::setAssetCacheKey() impl.

Crisspl committed:
asset-pipeline: OBJ loader adjustments part1

Crisspl committed:
asset-pipeline: OBJ loader adjustments part2

Crisspl committed:
asset-pipeline: Small IAsset changes

Crisspl committed:
ICPUTexture

Remove IAssetCreator

Crisspl committed:
asset-pipeline: Some fixes after CR

Crisspl committed:
asset-pipeline: cache aliasing related fixes

Crisspl committed:
asset-pipeline: BMP loader and ICPUTexture fixes

Also turned CImageData into asset

Crisspl committed:
object-cache: Added findAndStoreRange

TODO: Remove findRange from concurrent cache (this requires some changes in asset manager)

Crisspl committed:
object-cache: Added changeObjectKey()

+ got rid of macro defining findAndStoreRange()

Crisspl committed:
asset-pipeline: Fixed 'remove+insert as single op in changeAssetKey'; Added IAssetManager::dumpCachesDebug()

Crisspl committed:
object-cache: Fixed changeObjectKey()

Crisspl committed:
asset-pipeline: DDS loader (also made all those structs as loader's members)

+ mem-leak fix in BMP loader

Crisspl committed:
asset-pipeline: JPG loader

Crisspl committed:
asset-pipeline: PNG loader

Crisspl committed:
asset-pipeline: RGB loader

Crisspl committed:
asset-pipeline: TGA loader

Crisspl committed:
asset-pipeline: Changes to SAssetWriteParams (made root assset const), ICPUTexture, CImageData

Crisspl committed:
asset-pipeline: BMP writer

Crisspl committed:
asset-pipeline: JPG writer

Crisspl committed:
asset-pipeline: TGA writer

Crisspl committed:
asset-pipeline: PNG writer

Crisspl committed:
object-cache: Reworked findAndStoreRange(), added getAndStoreKeyRangeOrReserve()

Also: completely removed *begin()/*end() from concurrent cache and left only const overloads in regular cache

Crisspl committed:
asset-pipeline: Some small fixes

Crisspl committed:
ICPUTexture: many images representing one mipmap possible

Crisspl committed:
asset-pipeline: STL writer

Crisspl committed:
asset-pipeline: PLY writer

Crisspl committed:
object-cache: Added only value storing (not whole pair) findAndStoreRange()

Crisspl committed:
asset-pipeline: Updated findAssets()

Crisspl committed:
ICPUTexture: removed getLowestMip()

Crisspl committed:
asset-pipeline: Multiple fixes [kinda dirty commit]

[I|C]SceneManager dirty for testing/debugging purposes

Crisspl committed:
asset-pipeline: Image writers writes CImageData

Crisspl committed:
asset-pipeline: IAssetManager:: writer functions

Crisspl committed:
asset-pipeline: BAW loader - stack-based hierarchy level

Crisspl committed:
object-cache: findAndStoreRange() fix

Crisspl committed:
asset-pipeline: Making use of loader override

Crisspl committed:
asset-pipeline: make use of writer override

Crisspl committed:
asset-pipeline: X loader

Crisspl committed:
asset-pipeline: STL better validation

Crisspl committed:
ICPUTexture getMipMap() fix

Crisspl committed:
SMaterial -> SCPUMaterial, SGPUMaterial (temporary, unbuildable)

Crisspl committed:
Some small changes in ICPUTexture

Crisspl committed:
asset-pipeline: X loader with CPU textures

Crisspl committed:
asset-pipeline: OBJ loader with CPU textures

Crisspl committed:
asset-pipeline: BAW loader with CPU textures

Crisspl committed:
Changes in createGPUMeshFromCPU() due to CPU textures introduction

+ some dirty changes in CSceneManager
+ small changes in BAW writer due to CPU textures introduction

Crisspl committed:
asset-pipeline: Fixed image loaders

Crisspl committed:
Removed COgreMeshFileLoader and updated VS project file

Crisspl committed:
asset-pipeline: BAW writer, ICPUTexture::create() function

Crisspl committed:
asset-pipeline: Put asset manager and some fixes

Crisspl committed:
IDriver: cpu->gpu initial

Those can't be inline for sure at least as for now.. As i said - sick dependency hell

Crisspl committed:
cpu->gpu without Override

Crisspl committed:
cpu->gpu with IGPUObjectFromAssetConverter

Crisspl committed:
IGPUObjectFromAssetConverter, forgot to push this file

Crisspl committed:
Changes due to comments in prev commits

Changing directory structure in next commit

Crisspl committed:
Default converter instantied per call

Crisspl committed:
Reworked texture creation in CPUGPUconverter and removed texture-related functions from drivers

Crisspl committed:
Moved assets to asset namespace

Crisspl committed:
Changed file structure (moved asset-related things to include/irr/asset)

Takes into account your comments from f985d8f :)

Crisspl committed:
Color conversions

Without BC formats as for now

Crisspl committed:
BC decode functions

Crisspl committed:
colors: _scale param for SCALED only, <double> for FLOAT, NORM, SCALED only, uint64 for UINT only, int64 for INT only, removed previous ECF enums since it was mostly subset of Vulkan formats

Crisspl committed:
colors: fixed BC1 decode

Crisspl committed:
SColor.h seems like ready (x,y-based BC decode...)

Crisspl committed:
Find & replace ECFs over whole project

Crisspl committed:
color: Runtime decode/encode

todo: runtime convert(), planar formats, convert from BC

Crisspl committed:
color: Planar formats

Crisspl committed:
colors: Runtime-given formats convert

Crisspl committed:
colors: convert from BC

Crisspl committed:
colors: Divided color conversions into few headers

Crisspl committed:
color: Runtime convert functions actually needed to go to source file

msvc can't handle so template-heavy code but apparently moving it to separate translation unit solves the problem

Crisspl committed:
color: Added missing formats to runtime encode/decode

Crisspl committed:
colors: code for ASTC

Crisspl committed:
colors: decode ASTC

Crisspl committed:
Decode/encode functions generator code

Might be a little out of date in some places

Crisspl committed:
SRGB should be treated as normalized formats

Crisspl committed:
colors: SRGB should be treated as normalized

Crisspl committed:
Did a little workaround for Intellisense

Crisspl committed:
colors: Some more template magic for compile-time convertColor

Crisspl committed:
colors: Lil fix in convertColor

Crisspl committed:
colors: Updated ASTC

Crisspl committed:
colors: Fixed encodePixels a little

Crisspl committed:
oops

Crisspl committed:
Renamed ECOLOR_FORMAT to E_FORMAT

Crisspl committed:
Formats support query functions for OpenGL

Crisspl committed:
colors: Removing astc

Crisspl committed:
Added missing long unpushed files..

Crisspl committed:
Added glGetInternalFormativ to extension handler

Crisspl committed:
Better color format support (OpenGL)

Crisspl committed:
VS project file update

Crisspl committed:
Merge with buildaworld:master

Crisspl committed:
Found a bug that was f**king rgb10_a2 normals!

Crisspl committed:
Fixed lack of default writer override

Crisspl committed:
Updated meshloaders example

Crisspl committed:
Merge with Crisspl:asset (i left a few most invasive commits for 2nd merge)

Crisspl committed:
Updated meshloaders example once more

Crisspl committed:
getFormatChannelCount()

also deleted redundant files

Crisspl committed:
Migration to E_FORMAT for vertex attribs semi-finished

It is possible that we'll be encountering some easy fix bugs after this, however as for now [get|set]Attribute(), BAW writer/loader and mapping E_FORMAT->GLenum seems bugless
TODO:
* Test mesh optimizer
* Change baw version in writer and expected version value in BAW loader so that it won't do weird things while loading *.baw created before migration to E_FORMAT

Crisspl committed:
Got rid of lots of commented-out stuff

Crisspl committed:
Updated BAW format

as an extra: mesh optimizer with E_FORMAT seems to work fine, but maybe some more testing?

Crisspl committed:
This shouldnt be there

Crisspl committed:
Renamed map***Buffer functions to set***Buffer

Crisspl committed:
Number of small fixes

Crisspl committed:
Got rid of this format switch in vao

Crisspl committed:
More color format support fixes

Crisspl committed:
Mesh data descriptor: turned divisors into single uint32

Crisspl committed:
Fixed a few minor bugs

Crisspl committed:
bawformat: backward compatible loading

Crisspl committed:
Moved baw-blobs to asset ns

Crisspl committed:
Moved baw-blobs etc. to asset ns

Crisspl committed:
Updated CMake scripts (also fixes #193)

Crisspl committed:
Removed EF_nBIT_PIX formats

Crisspl committed:
CImage::getPixel: replaced format conversion with decode/encodePixels

Crisspl committed:
Added missing mapping E_FORMAT->GLenum of rgb10a2 texture formats

Crisspl committed:
Max buffer size query

Crisspl committed:
static_assert EBO-gurds for baw-blobs

Crisspl committed:
Moved CImageData to asset ns

Crisspl committed:
Moved CImageData.hto proper dir

Crisspl committed:
Moved IMeshBuffer<>, IMeshDataFormatDesc<> and ICPUMeshDataFormatDesc to asset ns

GPU equivalents coming to video ns in next commit

Crisspl committed:
Moved IGPUMeshDataFormatDesc and IGPUMeshBuffer to video ns

Crisspl committed:
Moved IGPUMeshBuffer.h to proper dir

Crisspl committed:
Moved IMesh<> to asset ns, moved IGPUMesh to video ns

plus divided into 2 separate files

Crisspl committed:
Moved SGPUMesh to video ns (and replaced its file)

Crisspl committed:
Moved [I|C]SkinnedMesh to video ns

Crisspl committed:
Renamed SCPUSkinMeshBuffer->ICPUSkinnedMeshBuffer (fixes #203)

Crisspl committed:
Moved E_FORMAT to asset ns

Crisspl committed:
Moved ALL loaders/writers to asset ns

Crisspl committed:
merge with builaworldnet:master

Crisspl committed:
Little corrections after merge

Crisspl committed:
Changed strides to uint32

Crisspl committed:
Reworked CMemoryFile and fixed a little baw loader

baw1 will probably NOT always be smaller than baw0

Crisspl committed:
N CPU -> 1 GPU bufs

Crisspl committed:
SOffsetBufferPair: protected dtor

Crisspl committed:
Number of smaller corrections

LoaderOverride got function for choosing relevant asset from found ones
ICPUMeshDataFormatDesc is now an asset
baw0->baw1: converted blobs are written straight to MemoryFile
getMappedBufferOffset/Stride no longer return reference

Crisspl committed:
CObjectCache::changeObjectKey doesnt use greet/dispose any more

Crisspl committed:
Moved GeometryCreator to asset ns

Crisspl committed:
Moved IAssetManager.cpp

Crisspl committed:
core::RAIIBasedExiter

Crisspl committed:
Moved IMeshManipulator to asset ns

Crisspl committed:
Revert due to weird behavior under release

Crisspl committed:
override->getLoadFile() shouldnt be called by loaders

Crisspl committed:
Extra param for getAsset(IReadFile*)

Crisspl committed:
Merged those two IAssetLoaderOverride's methods and moved GeometryCreator and MeshManipulator getters to IAssetManager

Crisspl committed:
Made IAssetManager::getAssetInHierarchy() protected

Crisspl committed:
Got rid of IMeshLoader, IMeshWriter, IMeshCache, CMeshCache

Crisspl committed:
Brand new CMemoryReadFile

Crisspl committed:
Updated COpenGLDriver::isAllowedTextureFormat()

Ready to merge!

Crisspl committed:
Updated CImageWriterJPG associated extensions

Crisspl committed:
Added PVRTC formats to E_FORMAT

However our glext.h doesnt define GL_ values for this. Is that normal?

Crisspl committed:
Added null_allocator

Crisspl committed:
CNullAllocatorReadFile

Crisspl committed:
ICPUTexture mipchain validation and some minor changes

Crisspl committed:
isHardwareBlendableFormat()

Crisspl committed:
oops

devshgraphicsprogramming commented on include/irr/asset/ICPUTexture.h:
```diff
@@ -187,16 +175,7 @@ class ICPUTexture : public IAsset
-            }
-        }
-        m_colorFormat = fmt;
+        m_colorFormat = (*std::find_if(m_mipmaps.begin(), m_mipmaps.end(), [](CImageData* mip) { return mip->getColorFormat() != EF_UNKNOWN; }))->getColorFormat();
```
what about the check that all the mips that don't have format `EF_UNKNOWN`, must have the same format?

Crisspl replied:
Since constructor (and those init functions as well) is protected, construction code assumes that `create()` already make sure that input `core::vector<asset::CImageData*>` is perfectly valid

devshgraphicsprogramming replied:
ok so this sort of validation takes place somewhere else?

Crisspl replied:
take a look at `create_impl()` and `validateMipchain()`

Crisspl committed:
ICPUTexture: oops

Crisspl committed:
baw loader: Template-based format conversions

Crisspl committed:
Made use of LinearAddressAllocator here

Crisspl committed:
Corrected AVX matrix multiplication benchmark

Crisspl committed:
Some simple fixes/corrections

Crisspl committed:
CObjectCache core:: containers test

Crisspl committed:
merge

Crisspl committed:
CObjectCache std->core revert

nah, msvc still crashing
Not building at this point, because im getting error at
SubAllocatedDataBuffer.h:159 (cannot convert std::pair to IGPUBuffer*)

Crisspl committed:
Builds on MSVC now! also updated CMake scripts

Take a look at my changes in SubAllocatedDataBuffer.h and ContiguousPoolAddressAllocator.h and tell me whether they're ok

devshgraphicsprogramming commented on include/irr/video/SubAllocatedDataBuffer.h:
```diff
@@ -153,7 +153,7 @@ class SubAllocatedDataBuffer : public virtual core::IReferenceCounted
 
             IGPUBuffer* retval;
-            static_if<is_std_get_0_defined<decltype(allocation)>::value >([&](auto f){
+            static_if<true/*is_std_get_0_defined<decltype(allocation)>::value*/ >([&](auto f){ // <-- todo, apparently MSVC SFINAE bug, however i can't even reproduce this in simpler form
```
that will break, badly I have need for both in the current codebase

Also what MSVC version are you using? this compiles for us on windows.

Crisspl replied:
VS 2017 (15.9.3)
i'll try changing toolset version.
Also i'll have to update cmake because it generates projects for VS 2015 and maybe there's the differenece between our setup

Crisspl replied:
yea, that was it! With v141 toolset it builds just fine

Crisspl committed:
Irrlicht must be built with VS2017 (v141) toolset!

Crisspl committed:
Updated ex07 source

Crisspl committed:
A little changes to cmake

* adding all the header files to Irrlicht build project (development purposes only, doesnt affect actual build)
* examples/tools binaries are outputted to source dir (and not build dir)

Crisspl committed:
Got rid of NASTY_OPENSSL_WORKAROUND

Crisspl committed:
Updated all the examples

Crisspl committed:
memleak fix

Crisspl committed:
Fixing even more memleaks!

Crisspl committed:
Dirty workaround of something

devshgraphicsprogramming commented on src/irr/asset/CBAWMeshFileLoader.h:
```diff
@@ -216,7 +216,7 @@ class CBAWMeshFileLoader : public asset::IAssetLoader
 
     template<uint64_t IntoVersion>
-    std::enable_if_t<IntoVersion, bool> formatConversionProlog(
+    std::enable_if_t<IntoVersion/*cannot convert into version 0*/, bool> formatConversionProlog(
```
isn't `std::enable_if_t`c++17?

Crisspl replied:
Nah, all this `_v` and `_t` from `<type_traits>` is c++14. And i'm always checking the standard version of a feature if i'm not sure, so don't worry :)

Crisspl committed:
Got rid of all memleaks

NOW IT'S PERFECT

Crisspl committed:
Merge branch 'master' of https://github.com/buildaworldnet/IrrlichtBAW

Crisspl committed:
Little fix due to addr alocator api change
